domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -1596,91 +1595,210 @@ public TotaledResults(ResultSet set, Integer total) {
-                replyHandler.handle(Future.failedFuture(e));
-              }
-            }).endHandler(v2 -> {
-              if (!transactionMode) {
-                connection.close();
-              }
-              replyHandler.handle(Future.succeededFuture());
-            });
+    connection.queryStream(selectQuery, stream -> {
+      if (stream.failed()) {
+        log.error(stream.cause().getMessage(), stream.cause());
+        replyHandler.handle(Future.failedFuture(stream.cause()));
+        return;
+      }
+      PostgresClientStreamResult<T> streamResult = new PostgresClientStreamResult(resultInfo);
 
+      SQLRowStream sqlRowStream = stream.result();
+      Promise<PostgresClientStreamResult<T>> promise = Promise.promise();
+      ResultsHelper<T> resultsHelper = new ResultsHelper<>(sqlRowStream, clazz);
+      boolean isAuditFlavored = isAuditFlavored(resultsHelper.clazz);
+      Map<String, Method> externalColumnSetters = getExternalColumnSetters(resultsHelper.columnNames,
+        resultsHelper.clazz, isAuditFlavored);
+      // fetch rows and produce ResultInfo when facets are all read
+      sqlRowStream.resultSetClosedHandler(v -> sqlRowStream.moreResults()).handler(r -> {
+        try {
+          JsonObject row = convertRowStreamArrayToObject(sqlRowStream, r);
+          T objRow = null;
+          // deserializeRow can not determine if count or user object
+          // in case where user T=Object
+          // skip the initial count result when facets are in use
+          if (resultsHelper.offset == 0 && !facets.isEmpty()) {
+            resultsHelper.facet = true;
           } else {
-            if (!transactionMode) {
-              connection.close();
+            objRow = (T) deserializeRow(resultsHelper, externalColumnSetters, isAuditFlavored, row);
+          }
+          if (!resultsHelper.facet) {
+            if (!promise.future().isComplete()) { // end of facets (if any) .. produce result
+              resultsHelper.facets.forEach((k, v) -> resultInfo.getFacets().add(v));
+              promise.complete(streamResult);
+              replyHandler.handle(promise.future());
             }
-            log.error(stream.cause().getMessage(), stream.cause());
-            replyHandler.handle(Future.failedFuture(stream.cause()));
+            streamResult.fireHandler(objRow);
           }
-        });
-
-      } catch (Exception e) {
-        if (!transactionMode) {
-          connection.close();
+          resultsHelper.offset++;
+        } catch (Exception e) {
+          if (!promise.future().isComplete()) {
+            promise.complete(streamResult);
+            replyHandler.handle(promise.future());
+          }
+          sqlRowStream.close();
+          log.error(e.getMessage(), e);
+          streamResult.fireExceptionHandler(e);
         }
-        log.error(e.getMessage(), e);
-        replyHandler.handle(Future.failedFuture(e));
-      }
+      }).endHandler(v2 -> {
```

Comment:
Can you add .exceptionHandler(e -> ... ?

Reply:
Good catch!